File: Linked_List/GFG_Remove duplicates from an unsorted linked list.py
Approach-1(hashset)

class Solution:
    #Function to remove duplicates from unsorted linked list.
    def removeDuplicates(self, head):
        # code here
        # return head after editing list
        Hashset=set()
        prev,itr=head,head
        
        while itr!=None:
            if itr.data in Hashset:
                prev.next=itr.next
            else:
                Hashset.add(itr.data)
                prev=itr
            itr=itr.next
        return head


Approach-2(two pointer)


def removeDuplicates(self,head):   
        ptr1 = None
        ptr2 = None
        dup = None
        ptr1 = head

		while (ptr1 != None and ptr1.next != None):
		             
		            ptr2 = ptr1
		 
		                # Compare the picked element with rest
		                # of the elements
            while (ptr2.next != None):
                 
                # If duplicate then delete it
                if (ptr1.data == ptr2.next.data):
                         
                    # Sequence of steps is important here
                    dup = ptr2.next
                    ptr2.next = ptr2.next.next
                else:
                    ptr2 = ptr2.next
                     
            ptr1 = ptr1.next
             
        return head

File: Linked_List/Java/GFG. Remove duplicates from an unsorted linked list.java
*******************************Optimal Solution*********************************8
TC-->O(N)
SC-->O(N)-->Hashing


class Solution
{
    //Function to remove duplicates from unsorted linked list.
    public Node removeDuplicates(Node head) 
    {
         // Your code here
         Node curr=head;
         Node prev=head;
         
         if(curr==null || curr.next==null) return curr;
         
         HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
         
         map.put(curr.data,1);
         
         curr=curr.next;
         
         while(curr!=null)
         {
             if(map.containsKey(curr.data))
             {
                 prev.next=curr.next;
                 curr=curr.next;
             }
             else{
                 
                 map.put(curr.data,1);
                 prev=prev.next;
                 curr=curr.next;
             }
 
         }
 
        return head;        
         
    }
}

